extensions/positron-python/python_files/posit/positron/patch/haystack.py

```diff
@@ -0,0 +1,70 @@
+#
+# Copyright (C) 2024 Posit Software, PBC. All rights reserved.
+# Licensed under the Elastic License 2.0. See LICENSE.txt for license information.
+#
+
+"""
+Patch for haystack-ai library to make it work correctly in Positron notebooks.
+
+The haystack-ai library has a function called is_in_jupyter() that checks if the code
+is running in a Jupyter notebook. However, it doesn't recognize Positron's custom IPython shell
+(PositronShell) as a Jupyter environment, causing compatibility issues.
+
+This patch overrides the function to always return True when run in Positron.
+"""
+
+import logging
+import importlib.util
+
+logger = logging.getLogger(__name__)
+
+
+def _patched_is_in_jupyter() -> bool:
+    """Replacement for haystack's is_in_jupyter that always returns True."""
+    return True
+
+
+def patch_haystack_is_in_jupyter() -> None:
+    """
+    Patch the haystack-ai library's is_in_jupyter function to return True in Positron.
+
+    This ensures haystack-ai correctly identifies Positron notebooks as Jupyter environments.
+    """
+    try:
+        # Check if haystack_ai is installed
+        if importlib.util.find_spec("haystack_ai") is None:
```

Comment:
is there a big speed difference that we need these 2 checks if we're going to try/except for import errors anyway?

Reply:
Not at all. Actually the whole implementation was inefficient.